knimin/lib/data_access.py

```diff
@@ -2450,6 +2452,866 @@ def set_deposited_ebi(self):
+                sql = "DELETE FROM pm.sample WHERE sample_id IN %s"
+                TRN.add(sql, [tuple(samples)])
+
+            # Delete the study
+            sql = "DELETE FROM pm.study WHERE study_id = %s"
+            TRN.add(sql, [study_id])
+            TRN.execute()
+
+    def set_study_samples(self, study_id, samples):
+        """Sets the samples for a study
+
+        Parameters
+        ----------
+        study_id : int
+            The study id
+        samples : iterable of str
+            The samples to be associated with the study
+
+        Raises
+        ------
+        ValueError
+            If the study ID does not exist
+        ValueError
+            If a sample is being removed and it was already plated
+        """
+        with TRN:
+            self._study_exists(study_id)
+            # Check if there are any samples that were already plated that
+            # are being removed
+            plated = set(chain.from_iterable(
+                self.get_study_plated_samples(study_id).values()))
+            samples = set(samples)
+            removed = plated - samples
+            if removed:
+                raise ValueError(
+                    "Can't remove samples from study %s. Offending samples: %s"
+                    % (study_id, ', '.join(removed)))
+
+            old_samples = set(self.get_study_samples(study_id))
+            new_samples = samples - old_samples
+            to_remove = old_samples - samples
+            if new_samples:
+                sql = "INSERT INTO pm.sample (sample_id) VALUES (%s)"
+                sql_args = [[s] for s in new_samples]
+                TRN.add(sql, sql_args, many=True)
+
+                sql = """INSERT INTO pm.study_sample (study_id, sample_id)
+                         VALUES (%s, %s)"""
+                sql_args = [[study_id, s] for s in new_samples]
+                TRN.add(sql, sql_args, many=True)
+                TRN.execute()
+            if to_remove:
+                to_remove = tuple(to_remove)
+                sql = """DELETE FROM pm.study_sample
+                         WHERE study_id = %s AND sample_id IN %s"""
+                TRN.add(sql, [study_id, to_remove])
+                sql = "DELETE FROM pm.sample WHERE sample_id IN %s"
+                TRN.add(sql, [to_remove])
+                TRN.execute()
+
+    def get_study_plated_samples(self, study_id):
+        """Gets the plated samples for the given study
+
+        Parameters
+        ----------
+        study_id : int
+            The study id
+
+        Returns
+        -------
+        dict {plate_id: samples}
+            The plated samples grouped by plate id
+
+        Raises
+        ------
+        ValueError
+            If the study ID does not exist
+        """
+        with TRN:
+            self._study_exists(study_id)
+            sql = """SELECT sample_plate_id,
+                            ARRAY_AGG(DISTINCT sample_id ORDER BY sample_id)
```

Comment:
@sjanssen2 and I saw some errors with these types of aggregations associated with postgres version. I dont know if this query is impacted, just a general FYI. I believe we're a bit dated on what is up server-side and it may differ from what is being run locally on dev systems.